docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+time than before.
+
+## Background
+
+At present, the optimization procedure of the planner is separated into two
+phases. The first phase, namely the "Logical Optimization", only applies the
+rules which always beneficial. The second phase, which is called the "Physical
+Optimization", takes the cost of different physical operator implementations
+into consideration and chooses the best physical plan with the lowest cost.
+
+However, there are some other transformations which are not always beneficial
+for all the scenarios. For example, aggregate push down, aggregate pull up, in
+subquery unfold, etc.
+
+Another drawback of the current planner is the poor extensibility. It's hard to
+add a new rule even if it's beneficial for all the scenarios: we have to
+consider the order of diffenent optimization rules carefully.
+
+The physical optimization for the operators on the storage layer also suffers
+from the poor extensibility. In the present planner, we use "root" and "cop"
+task to distinguish the operators executed on TiDB and the storage layer, TiKV.
+The way to seperate a "cop" task is also  not extensible, and "cop" tasks are
+highly tied with "root" task. For exmple, we can only has a `Stream` aggregate
+operator in the "cop" task if the "root" task also is a `Stream` aggregate
+operator.
+
+
+## Glossary
+
+- **Expression**
+
+  In this proposal, expression is used to specify a logical plan. Expression can
+  be expressed to a tree-like structure, the child of an expression is also an
+  expression.
+
+- **Expression Group**(or **Group**)
+
+  Expression group is used to store all the logically equivalent expressions.
+  It's a set of **Group Expressions**.
+
+- **Group Expression**
+
+  Group expression is used to store all the logically equivalent expressions
+  which have the same root operator. Different from a normal expression, the
+  children of a group expression are expression **Groups**, not expressions.
+  Another property of group expression is that the child group references is
+  never changed once the group expression is created.
+
+  With the concept of **Group** and **Group Expression**, all the logically
+  equivalent expressions can be stored in a root Group.
+
+- **Transformation Rule**
+
+  The transform rule is used to transform a logical plan to another equivalent
+  logical plan. It is used to explore all the logically equivalent query plans
+  belonging to an expression group.
+
+- **Implementation Rule**
+
+  The implementation rule is used to implement a logical expression operator to
+  a physical operator. For example, with implementation rules, a logical `Join`
+  operator can be implementated to `HashJoin`/`MergeJoin`/`IndexJoin`, etc.
+
+- **Enforcing Rule**
+
+  Rule to construct an enforcer to satisfy the required physical properties.
+
+- **Pattern**
+
+  Pattern describes a piece of a logical expression. It's a tree-like structure,
```

Comment:
What's the different between `Pattern` and `Expression`?  `Expression` is also a tree-like structure.

Reply:
`Pattern` is to describe a pattern of `Expression` tree. It only concerns the type of `Expression` node.